endpoints/put.go

```diff
@@ -197,74 +205,106 @@ func (e *PutHandler) handle(w http.ResponseWriter, r *http.Request, ps httproute
+			return resp.err
 		}
+	}
 
-		// Only allow setting a provided key if configured (and ensure a key is provided).
-		if e.cfg.allowKeys && len(p.Key) > 0 {
-			resps.Responses[i].UUID = p.Key
-			e.metrics.RecordPutKeyProvided()
-		} else if resps.Responses[i].UUID, err = utils.GenerateRandomID(); err != nil {
-			return utils.NewPBCError(utils.PUT_INTERNAL_SERVER, "Error generating version 4 UUID")
+	return nil
+}
+
+// put parses the putObject, validates it and calls the back-end storage Put() function this Prebid Cache instance
+// is using. Returns a putResponseObject storing either the corresponding UUID's data was stored under, or an error
+// if any.
+func (e *PutHandler) put(po *putObject, resp *putResponseObject, index int, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	toCache, err := parsePutObject(*po)
+	if err != nil {
+		resp.err = err
+	}
+
+	// Only allow setting a provided key if configured (and ensure a key is provided).
+	if e.cfg.allowKeys && len(po.Key) > 0 {
+		// put object comes with custom key, which we are allowed to use
+		resp.UUID = po.Key
+		e.metrics.RecordPutKeyProvided()
+	} else {
+		// Either put object doesn't come with a custom key or Prebid Cache is configured
+		// to not use custom keys. Generate a random UUID
+		if resp.UUID, err = utils.GenerateRandomID(); err != nil {
+			resp.UUID = ""
+			resp.err = utils.NewPBCError(utils.PUT_INTERNAL_SERVER, "Error generating version 4 UUID")
+			return
 		}
+	}
+
+	// If we have a blank UUID, don't store anything.
+	// Eventually we may want to provide error details, but as of today this is the only non-fatal error
+	// Future error details could go into a second property of the Responses object, such as "errors"
+	if len(resp.UUID) > 0 {
+		ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
+		defer cancel()
 
-		// If we have a blank UUID, don't store anything.
-		// Eventually we may want to provide error details, but as of today this is the only non-fatal error
-		// Future error details could go into a second property of the Responses object, such as "errors"
-		if len(resps.Responses[i].UUID) > 0 {
-			ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
-			defer cancel()
-
-			err = e.backend.Put(ctx, resps.Responses[i].UUID, toCache, p.TTLSeconds)
-			if err != nil {
-				if pbcErr, isPbcErr := err.(utils.PBCError); isPbcErr && pbcErr.Type == utils.RECORD_EXISTS {
-					// Record didn't get overwritten, return a reponse with an empty UUID string
-					resps.Responses[i].UUID = ""
-				} else {
-					return logBackendError(err, i)
-				}
+		err = e.backend.Put(ctx, resp.UUID, toCache, po.TTLSeconds)
+		if err != nil {
+			if pbcErr, isPbcErr := err.(utils.PBCError); isPbcErr && pbcErr.Type == utils.RECORD_EXISTS {
+				// Record didn't get overwritten, return a response with an empty UUID string
+				resp.UUID = ""
+			} else {
+				resp.err = classifyBackendError(err, index)
 			}
```

Comment:
If we receive a record exists error when attempting to put via `e.backend.Put`, shouldn't we also set `resp.err` in this case? This currently only happens when it is not a `RECORD_EXISTS` error due to the `else` here.

Reply:
That's Prebid Cache's current behavior as noted in the `README.md`. :
```
 #### Overwritting values
 
 Prebid Cache does not allow overwritting any value, for either autogenerated or custom keys. If an entry already exists for a given key, it will not be overwitten, and an empty string will be returned as the `uuid` value of that entry.....
README.md
```
In the near future we might want to provide error details. Probably inside the `"errors"` property of the `Responses` objects.